src/group.cpp

```diff
@@ -303,20 +304,24 @@ void Group::assign(int narg, char **arg)
       else if (category == MOLECULE) tattribute = atom->molecule;
       else if (category == ID) tattribute = atom->tag;
 
-      char *ptr;
       tagint start,stop,delta;
 
       for (int iarg = 2; iarg < narg; iarg++) {
         delta = 1;
-        if (strchr(arg[iarg],':')) {
-          ptr = strtok(arg[iarg],":");
-          start = utils::tnumeric(FLERR,ptr,false,lmp);
-          ptr = strtok(nullptr,":");
-          stop = utils::tnumeric(FLERR,ptr,false,lmp);
-          ptr = strtok(nullptr,":");
-          if (ptr) delta = utils::tnumeric(FLERR,ptr,false,lmp);
-        } else {
-          start = stop = utils::tnumeric(FLERR,arg[iarg],false,lmp);
+        try {
+          ValueTokenizer values(arg[iarg],":");
+          start = values.next_tagint();
+          if (utils::strmatch(arg[iarg],"^\\d+$")) {
```

Comment:
@akohlmey Instead of double-parsing with `strmatch` here you could instead store `values.count()` as `ncount` and then do:
```
if (ncount == 1) {
  ...
} else if (ncount == 2) {
  ...
} else if (ncount == 3) {
  ...
}
```
or use a `switch` statement. The `next_tagint()` functions will already throw a `TokenizerException` if the value isn't an integer.

Reply:
I had something similar at some point but it was not enforcing the syntax properly. Incorrect input like "1:" or "1::4" or ":2" are happily accepted while they are not allowed. I added the `utils::strmatch()` calls to enforce the syntax more strictly. The counting of tokens is then a welcome side effect as I can omit `if (values.has_next())`.